### ISSUE TITLE
Unify syntax for watching providers

### ISSUE THREAD

**Issue by TimWhiting** [rrousselGit/river_pod#328]
This was a thought that came to me recently.

It might be a little controversial, but I wanted to get other opinions / see other sides of the issue. Especially from your viewpoint @rrousselGit 

`StateNotifierProvider` is the only provider that has two different "aspects" of the state. One is the actual `.state` of the `StateNotifier`, and one is the `StateNotifier` itself. One is the primary state and the other is just an interface for `mutating/updating` the state. The one that is commonly `watch`ed is the state, and the one that should be primarily `read` is the notifier.

Common usage (in my experience) dictates that the:
- notifier 
  - should be `context.read` in callbacks in the widget tree
  - should also be `ref.read` in any providers wanting to use the interface (not in the provider create function body, but a `ref.read` - aka `Reader` - should be passed into the constructor of the class / repository, to allow it access to the interface).
- state
  - should be `watch` or `ref.watch`ed in most cases.
 
Also, it seems like there is a lot of confusion in the issues about why `StateNotifer` is "different" than the other providers when it comes to `watch`ing the state in a widget. I know that there really isn't any difference because the `StateNotifier` subclass is the thing that is being provided. But it is interesting in that it is the ?only? provider that has that different syntax for reading the state. I guess that isn't completely true given the fact that you can read/watch? a `FutureProvider`'s `.future`. However, there is this disconnect in the fact that the common thing to be `watch`ed in a `StateNotifierProvider` also has a "different" API.

Obviously the current proposed solution is to just document these aspects / differences better. If so, I'm willing to spend some time working on some documentation as requested in #326 and various other issues. However, I'm wondering if there is a better API specifically to unify syntax for watching the state of a StateNotifier (which is the most common thing that should be watched if my experience is correct). 

Again, this is based off of my experience which might be different than others' experience. I'm not proposing any specific solution here, just making an observation, and want to understand the viewpoints of others.
 
Specifically I want to point out that the current API makes sense to me and others who frequently use this library and understand how it works, but to beginners and newcomers, it is strange, which is why it comes up so frequently in the issue tracker. I'm guessing your devtool / lint package will issue warnings to enforce / warn about the common usage mentioned previously.

**Comment by rrousselGit** [comment]
> `StateNotifierProvider` is the only provider that has two different "aspects" of the state

As you've said, later, it isn't the only one

StreamProvider/FutureProvider do it too, with:

```dart
StreamProvider<int> provider;

AsyncValue<int> asyncValue = ref.watch(provider);
Future<int> lastValue = ref.watch(provider.last);
Stream<int> stream = ref.watch(provider.stream);
```



Personally, what I'd like is to be able to switch the behaviour of `ref.watch(stateNotifierProvider)` and `ref.watch(stateNotifierProvider.state)`

The `.state` should be the default behaviour IMO. The problem is... We can't with Dart at the moment.

Supporting it would required changing `StateNotifierProvider<MyNotifier>` into `StateNotifierProvider<MyNotifer, MyState>` which is quite verbose and redundant. Hence why it is "backward" at the moment.

**Comment by rrousselGit** [comment]
One solution though could be to:

- remove the `@protected` on `StateNotifier.state`
- have `watch(stateNotifierProvider)` "watch" the state
- make people do `final state = watch(provider).state` instead of `final state = watch(provider.state)`
- expose a `final notifier = watch(provider.notifier)` that does **not** listen to the state
- do the same with ChangeNotifierProvider which could benefit from such `.notifier` too.

**Comment by TimWhiting** [comment]
Personally I think your last solution is what a lot of people are looking for, and mimics the `StateProvider`'s
`watch(provider).state`

**Comment by rrousselGit** [comment]
What I fail to understand is: Why do newcomers write `final state = watch(provider).state`? 

Because the IDE gives a very clear warning: "state is protected"

It seems like newcomers ignore the warning but don't make the connection between the warning and their UI not rebuilding.

**Comment by rrousselGit** [comment]
> Personally I think your last solution is what a lot of people are looking for, and mimics the `StateProvider`'s
> `watch(provider).state`

The biggest issue with that is, the current separation exists to force user to not have any state in their `StateNotifier` _outside of the `state` property`.



With this change, people could write:


```dart
class Notifier extend StateNotifier<void> {
  int _count = 0;
  int get count => _count;
}


final count = ref.watch(notifier).count;
```

**Comment by TimWhiting** [comment]
> Because the IDE gives a very clear warning: "state is protected"

Agreed, the IDE is rather clear already as far as it being protected, but they might assume if `watch(provider).state` is protected, than `provider.state` might also be protected. Or I think the more common thing is that they don't see / understand the warning since it mentions nothing about the right way to do it, and they don't know about the `.state` on the `provider`.

**Comment by TimWhiting** [comment]
> The biggest issue with that is, the current separation exists to force user to not have any state in their `StateNotifier` _outside of the `state` property_.
> 
> With this change, people could write:
> 
> ```dart
> class Notifier extend StateNotifier<void> {
>   int _count = 0;
>   int get count => _count;
> }
> 
> 
> final count = ref.watch(notifier).count;
> ```
> 
> And that's already a mistake newcomers do. This change would reinforce it

I can see that this would happen. In the case of `ChangeNotifier` this misunderstanding can also happen, but it is clear that you have to call `notifyListeners`. But maybe a lint on `StateNotifier` subclasses to not declare any public accessible fields or getters would be a solution to that problem.

**Comment by rrousselGit** [comment]
> But maybe a lint on StateNotifier subclasses to not declare any public accessible fields or getters would be a solution to that problem.

But then we can have a lint to improve the error message on `watch(provider).state`

Maybe a first step would be to better document that.

Something like:

```
My Widget is not rebuilding when my StateNotifier changes, why?

do watch(provider.state) instead of watch(provider).state
```

**Comment by TimWhiting** [comment]
> But then we can have a lint to improve the error message on `watch(provider).state`

Understood, but it doesn't change the fact that the most common thing to `watch` on a `StateNotifier` is the `.state`. Which means when you want to `watch` a provider you have to remember if the `provider` you are watching is a `StateNotifier` or not. Of course autocomplete and intellisense helps with that.

**Comment by rrousselGit** [comment]
Another possibility is to have an assert in debug mode that prevents reading .state inside ConsumerWidget.build
\

**Comment by TimWhiting** [comment]
^ This might work. Would it also work for `useProvider`, I assume so.

**Comment by rrousselGit** [comment]
> On the other side of things, depending on debug mode means that the developer has to potentially touch every screen / function of the app to get this error. (i.e. it is not compile safe)?

That shouldn't be necessary, because the `@protected` still exists.

The main issue is that newcomers ignore it. But for people writing tests and caring about compile-safety, they'd have their CI fail if there's a warning

**Comment by wujek-srujek** [comment]
> What I fail to understand is: Why do newcomers write `final state = watch(provider).state`?
> 
> Because the IDE gives a very clear warning: "state is protected"
> 
> It seems like newcomers ignore the warning but don't make the connection between the warning and their UI not rebuilding.

I'm a newcomer and I don't know any better and am pretty confused. For example, this snippet here:
https://riverpod.dev/docs/concepts/reading#consumer
uses `final state = watch(provider).state`. It means that the syntax is different for a `Provider` and `StateNotifierProvider`. I did notice the warning, but then I checked what is done in `StateController`, found this:
```dart
// Remove the protected status
  @override
  T get state => super.state;
```
and did the same for my class. The fact that both `watch(aProvider.state)` and `watch(aProvider).state` compile (a warning is just this, a warning) doesn't help at all.

**Comment by wujek-srujek** [comment]
I think that these two snippets will cause a lot of bugs and confusion, the 'incorrect' syntax should ideally be a compilation error, if possible. Lints are just this, lints, warnings - don't forget not everyone is as good a developer as you guys (I work on a project where many teams don't even use the analyzer...), but they still want to get the job done, and you probably also want them to get it done using your package. The idea @rrousselGit had about an assertion for `watch(provider).state` is kind of contradicting to one of the main tenets of Riverpod - compile time safety. In this case, the code will compile, but I will get a runtime exception at most, not ideal.

**Comment by rrousselGit** [comment]
The problem is, this `@protected` warning has more uses than just avoiding this bug. Removing it has undesired side-effects.

This `@protected` also promotes a centralised business logic. Removing it also allows:

```dart
class Counter extends StateNotifier<int> {
  Counter(): super(0);
}

final provider = StateNotifierProvider<Counter>((ref) => Counter());

RaisedButton(
  onPressed: () => 
    // The logic shouldn't be here. Instead this ++ logic should be in a method inside `Counter`
    context.read(provider).state++,
)
```

**Comment by wujek-srujek** [comment]
Also, the syntax `watch(aProvider.state)` depends on an extensions in `base.dart`, so someone who knows `StateNotifier` doesn't necessarily expect it to suddenly have a `state` property. When I command-click the `state` property, I land here:
```dart
/// Adds [state] to [StateNotifierProvider.autoDispose].
extension StateNotifierStateProviderX<Value>
    on StateNotifierProvider<StateNotifier<Value>> {
  /// {@macro riverpod.statenotifierprovider.state.provider}
  StateNotifierStateProvider<Value> get state {
    _state ??= StateNotifierStateProvider<Value>._(this);
    return _state as StateNotifierStateProvider<Value>;
  }
}
```
Why is my provider suddenly an 'autoDispose' one? Or is it just an IDE quirk?

**Comment by TimWhiting** [comment]
@rrousselGit 
There are extremely rare, but valid use cases for ignoring a `@protected` annotation. Personally I feel like that is why the dart team made it an annotation rather than a language feature. But that is not the point of this issue.

Back to the topic of the issue. I feel like I'm not the only one who would prefer the alternative 
* have `watch(stateNotifierProvider)` "watch" the state
* make people do `final state = watch(provider).state` instead of `final state = watch(provider.state)`
* expose a `final notifier = watch(provider.notifier)` that does not listen to the state
* do the same with `ChangeNotifierProvider` which could benefit from such `.notifier` too.

Personally I think you could do this without having removing the `@protected` annotation from `StateNotifier.state` 
by returning a new object.
For example:
```dart
class StateNotifierWatchedState<T> {
  const StateNotifierWatchedState(this.state);
  final T state;
}
```
And have `StateNotifierWatchedState<T>` be the return type for `watch(someStateNotifierProvider<T>);`

What do you think?

**Comment by rrousselGit** [comment]
I don't see the benefit of such object.

If we're really going for that, might as well update `StateNotifierProvider` to:

```dart
class MyNotifier extends StateNotifier<int> {...}

final provider = StateNotifierProvider<MyNotifier, int>(...)

int state = watch(provider);
MyNotifier notifier = watch(provider.notifier);
```


The inconvenience of having to specify a redundant generic parameter is much lower than that of having to specify `.state` all over the place

**Comment by rrousselGit** [comment]
Thinking about it, a case could be made that the inconvenience of having to write `watch(provider.state)` is higher than having a redundant generic argument...

Maybe I should just add that extra generic parameter and have the expected default behaviour.

Then when Dart evolves, we'll be able to remove that generic parameter.

**Comment by wujek-srujek** [comment]
Can this generic parameter be inferred or will it be necessary to always specify it explicitly?

**Comment by rrousselGit** [comment]
It cannot be inferred. Which is one of the main reasons why I used this workaround.

**Comment by SteveAlexander** [comment]
I guess this is the issue to watch: https://github.com/dart-lang/language/issues/102

**Comment by rrousselGit** [comment]
I think this can be closed now that we have #341
